### ISSUE TITLE
Sony HDR-TD10 support

### ISSUE THREAD

**Issue by rsr-lpg** [ma1co/Sony-PMCA-RE#212]
Hello,
Would it be possible to add HDR-TD10 camera to the list? Basically, I need the language unlock tweak for Japanese cameras.
Thank you very much in advance.

**Comment by ma1co** [comment]
I don't think you tried with the correct .zip file. Can you try this one instead? [0x11010000.zip](https://github.com/ma1co/Sony-PMCA-RE/files/5437971/0x11010000.zip)

The TD10 is very interesting, because it contains two image processors. The main USB port on the side connects to the main processor.
On the bottom of the camera, there should be a hidden flap next to the battery ejector, with an additional USB port connecting to the secondary processor. I don't think we need that for the language unlock (I suppose that the main processor displays the GUI), but it might be interesting to dump the firmware of the secondary processor, too.

**Comment by rsr-lpg** [comment]
Thank you for the fast reply.
I used the provided file and got the following result:
```
Using drivers Windows-MSC, Windows-MTP
Looking for Sony devices

Querying mass storage device
Sony Camcorder is a camera in mass storage mode

Getting device info
Model name: HDR-TD10

Trying 100 firmware images
Firmware update error: Invalid version
Success: Found matching file: gen2_0x11010011_0.dat
```
Regarding the 2nd USB port, I found it at the location you described, so I guess it's possible to get a dump of the secondary processor. Please let me know further instructions on how to proceed.

**Comment by rsr-lpg** [comment]
I've just tried your suggestion. Here is the response:
```
Using drivers Windows-MSC, Windows-MTP
Looking for Sony devices
No devices found. Ensure your camera is connected.
```
I mention that there is no feedback on the camera display after connecting to the second USB port. Is there an engineering menu I need to access in order to activate this USB port?

**Comment by ma1co** [comment]
Ok, let's forget about the second USB port for now. I would expect that there is a way to update the firmware of the secondary processor somehow, but I have no idea how to activate the port.

You should now be able to dump the (main) firmware.

* Update to the newest [development build](https://ci.appveyor.com/project/ma1co/sony-pmca-re/build/artifacts) of pmca-console.
* Connect the camera.
* Run `pmca-console updatershell`
* Run the following updatershell command to dump the firmware (this will take some time to complete): `pull /dev/nflasha`
* Run the following updatershell command to dump the bootloader partitions: `bootloader`
* Run the following updatershell command to reboot your camera back to normal mode: `exit`
* You should now have files called `nflasha` and `boot` on your computer. Please zip them and upload the file to my [dropbox](https://www.dropbox.com/request/WXQ1IKIoPo4ECK8CFxRx).

Thank you for your help!

**Comment by rsr-lpg** [comment]
I've packed the files into hdr-td10_dump.zip and uploaded to your dropbox.

**Comment by rsr-lpg** [comment]
I've run the pmca-gui on my HDR-TD10 camera and managed to unlock all the languages. It worked like a charm 🥇 
I guess you can now include it in the list of supported models. I'd like to take the opportunity to thank you very much for your great support!
In case you have some new ideas on how to deal with the second USB port, please let me know, I would be more than happy to take a firmware dump from my camera and send it to you.

**Comment by ma1co** [comment]
Thank you for the dump! I'm glad it worked. If I have any idea regarding the second USB port, I will let you know.

**Comment by bg5bdt** [comment]
dear MR.
i'm chinese ,i want to change hdr-td10 unlock language,what can i do. please help.

**Comment by tinhtu2012** [comment]
@bg5bdt  Try that file 

[pmca-gui-v0.17-46-g2cdf126-win.zip](https://github.com/ma1co/Sony-PMCA-RE/files/6481507/pmca-gui-v0.17-46-g2cdf126-win.zip)

**Comment by 3DBrian** [comment]
I have a Japanese HDR-TD20 (Japan model) which I noticed hasn't been added to the list.

I'm trying to follow the directions to pull the firmware (having read the readme file) but I'm totally lost at the commands, etc.  Blame it on being a Mac user trying to use a PC.  I'm using the GUI exe from above, but it's not clear how/what/where I put the commands in (I tried the CMD but no luck).

If anyone can help, please let an old fool know and I'll dump the firmware so it can perhaps be updated to English/language unlock.  Appreciate any help!

**Comment by ma1co** [comment]
@3DBrian Here are more detailed instructions (I hope):

* Download the newest development build of pmca-console [here](https://ci.appveyor.com/project/ma1co/sony-pmca-re/build/artifacts) (not pmca-gui).
* Download [0x21010000.zip](https://github.com/ma1co/Sony-PMCA-RE/files/7998449/0x21010000.zip) and place it in the same folder (no need to unpack).
* Power on your camera and connect it over USB to your computer.
* Open a command prompt (search for it in the start menu) and change to the directory containing the two files: `cd C:\path\to\Downloads`
* Run the following command: `pmca-console-v0.17-70-g1958a45-win.exe guess_firmware -f 0x21010000.zip`
* With a bit of luck, the script should find the model id of your camera.

Let me know if you need any help.

**Comment by 3DBrian** [comment]
Thank you for the quick feedback.  So I did as you suggested, and it gave me the same info the GUI did.  I think where I'm getting lost is how to execute the program to actually download the firmware from the camera, specifically in these instructions that were posted above:

"You should now be able to dump the (main) firmware.

Update to the newest [development build](https://ci.appveyor.com/project/ma1co/sony-pmca-re/build/artifacts) of pmca-console.
Connect the camera.
Run pmca-console updatershell
Run the following updatershell command to dump the firmware (this will take some time to complete): pull /dev/nflasha
Run the following updatershell command to dump the bootloader partitions: bootloader
Run the following updatershell command to reboot your camera back to normal mode: exit
You should now have files called nflasha and boot on your computer. Please zip them and upload the file to my [dropbox](https://www.dropbox.com/request/WXQ1IKIoPo4ECK8CFxRx)."

I'm attaching the screenshots I took so maybe it's abundantly clear where/what I'm doing wrong.  I'm sure I'm leaving something out.



![IMG_7199](https://user-images.githubusercontent.com/98929627/152451973-e2b759dc-e25d-40f9-9421-8dcbe2a56e2c.jpeg)
![IMG_7198](https://user-images.githubusercontent.com/98929627/152451962-10b1b66a-8142-4955-8775-66c1ad1d1e43.jpeg)
![IMG_7196](https://user-images.githubusercontent.com/98929627/152451966-ffdbec2d-f8d8-412b-9d27-aeb13f47aef2.jpeg)


![IMG_7197](https://user-images.githubusercontent.com/98929627/152451950-f756aeca-10b1-4ea4-8eb9-dd61b36e7395.jpeg)

**Comment by ma1co** [comment]
@3DBrian The second screenshot is all I needed to add support for this camera. The next step is to dump the firmware:

* Update to the newest [development build](https://ci.appveyor.com/project/ma1co/sony-pmca-re/build/artifacts) of pmca-console.
* Connect the camera.
* Run `pmca-console-v0.17-71-gf0aacae-win.exe updatershell`
* Run the following updatershell command to dump the firmware (this will take some time to complete): `pull /dev/nflasha`
* Run the following updatershell command to dump the bootloader partitions: `bootloader`
* Run the following updatershell command to reboot your camera back to normal mode: `exit`
* You should now have files called `nflasha` and `boot` on your computer. Please zip them and upload the file to my [dropbox](https://www.dropbox.com/request/WXQ1IKIoPo4ECK8CFxRx).

Thank you for your help!

**Comment by 3DBrian** [comment]
Ok now it's clear why (you needed to update).  zip file using the camera's model with the two files in your dropbox

Thank you for the help!

**Comment by bg5bdt** [comment]
你好我现在在温州没有上！我会尽快回复你的！
有急事请打我电话联系！

**Comment by ma1co** [comment]
@3DBrian Thank you for the dump! It should be possible to tweak the languages using the [development builds](https://ci.appveyor.com/project/ma1co/sony-pmca-re/build/artifacts) of pmca-console / pmca-gui.

**Comment by 3DBrian** [comment]
Thanks, so I should just follow the instructions above for the Td10?  Or something else?

**Comment by ma1co** [comment]
@3DBrian The easiest way is to use the "Tweaks" tab in pmca-gui. Click "Start tweaking", then there should be an option to unlock all languages.

**Comment by 3DBrian** [comment]
> @3DBrian The easiest way is to use the "Tweaks" tab in pmca-gui. Click "Start tweaking", then there should be an option to unlock all languages.

This worked thank you!

Originally it would default back to Japanese, but I discovered in the menu turning “demo mode” to “off” enabled language settings to be saved.

thanks Again!

**Comment by 3DBrian** [comment]
@ma1co I just got another HDR-TD10 from Japan.  When I connect and use pmca-gui, after switching to MTP mode and query camera, there is japanese text which I accept.  Recognizes the camera.  When I go to "Tweaks" and select the options (unlock languages for example), and select "ok" it very quickly says "Done" and my camera screen is black.  But when I reboot the camera, there are no changes to language nor a way to access anything but Japanese.  What am I doing wrong?  Thanks!
